File: DAIRemoteApp/app/src/main/java/com/example/dairemote_app/KeyboardToolbar.java
package com.example.dairemote_app;

import android.view.View;
import android.widget.GridLayout;
import android.widget.TextView;

import androidx.appcompat.widget.Toolbar;

public class KeyboardToolbar {
    private Toolbar keyboardToolbar;
    private GridLayout keyboardExtraBtnsLayout;
    int modifier1, modifier2, modifier3, modifier4, modifier5, moreOpts;
    private final StringBuilder keyCombination = new StringBuilder();
    private boolean winActive, ctrlActive, shiftActive, altActive, fnActive, modifierToggled = false;
    private final String[] p1Keys = {"{INS}", "{DEL}", "{PRTSC}", "{TAB}", "{UP}", "{ESC}", "UP", "DOWN", "MUTE", "{LEFT}", "{DOWN}", "{RIGHT}"};
    private final String[] p2Keys = {"{F1}", "{F2}", "{F3}", "{F4}", "{F5}", "{F6}", "{F7}", "{F8}", "{F9}", "{F10}", "{F11}", "{F12}"};
    private TextView[] p1RowsButtons;
    private TextView[] p2RowsButtons;
    private TextView keyboardTextInputView;
    private int parenthesesCount = 0;
    private int currentPageIndex = 0;

    KeyboardToolbar(int moreopts, int modifier1, int modifier2, int modifier3, int modifier4, int modifier5, TextView textView, Toolbar toolbar, GridLayout toolbarLayout, TextView[] page1, TextView[] page2) {
        this.moreOpts = moreopts;
        this.modifier1 = modifier1;
        this.modifier2 = modifier2;
        this.modifier3 = modifier3;
        this.modifier4 = modifier4;
        this.modifier5 = modifier5;
        this.keyboardTextInputView = textView;
        this.keyboardToolbar = toolbar;
        this.keyboardExtraBtnsLayout = toolbarLayout;
        this.p1RowsButtons = page1;
        this.p2RowsButtons = page2;
    }

    public int GetParenthesesCount() {
        return this.parenthesesCount;
    }

    public Toolbar GetKeyboardToolbar() {
        return this.keyboardToolbar;
    }

    public GridLayout GeyKeyboardLayout() {
        return this.keyboardExtraBtnsLayout;
    }

    public boolean GetModifierToggled() {
        return this.modifierToggled;
    }

    public void SetModifierToggled(boolean toggled) {
        this.modifierToggled = toggled;
    }

    public String[] GetKeys(int page) {
        if(page == 0) {
            return p1Keys;
        }
        return p2Keys;
    }

    public TextView[] GetButtons(int page) {
        if(page == 0) {
            return p1RowsButtons;
        }
        return p2RowsButtons;
    }

    public int GetCurrentToolbarPage() {
        return this.currentPageIndex;
    }

    public void SetCurrentToolbarPage(int page) {
        this.currentPageIndex = page;
    }

    public int NextToolbarPage() {
        SetCurrentToolbarPage((GetCurrentToolbarPage() + 1) % 2);
        return GetCurrentToolbarPage();
    }

    public TextView GetKeyboardTextView() {
        return this.keyboardTextInputView;
    }

    public StringBuilder GetKeyCombination() {
        return this.keyCombination;
    }

    public void AppendKeyCombination(String character) {
        keyCombination.append(character);
    }

    public void AppendKeyCombination(char character) {
        keyCombination.append(character);
    }

    public void AddParentheses() {
        if (GetParenthesesCount() > 0) {
            for (int i = 0; i < parenthesesCount; i++) {
                AppendKeyCombination(")");
            }
            keyboardTextInputView.setText("");
        }
    }

    public void ToggleKeyboardToolbar(boolean open) {
        if (open) {
            if (GetKeyboardToolbar() != null && !(GetKeyboardToolbar().getVisibility() == View.VISIBLE)) {
                GetKeyboardToolbar().setVisibility(View.VISIBLE);
                GeyKeyboardLayout().setVisibility(View.VISIBLE);
                keyboardExtraSetRowVisibility(GetCurrentToolbarPage());
            }
        } else {
            if (GetKeyboardToolbar() != null && GetKeyboardToolbar().getVisibility() == View.VISIBLE) {
                GetKeyboardToolbar().setVisibility(View.GONE);
                GeyKeyboardLayout().setVisibility(View.GONE);
            }
        }
    }

    public void keyboardExtraSetRowVisibility(int pageIndex) {
        // Hide buttons for the current page
        for (TextView button : (pageIndex == 0) ? GetButtons(1):GetButtons(0)) {
            button.setVisibility(View.GONE);
        }

        // Show buttons for the current page
        for (TextView button : (pageIndex == 0) ? GetButtons(0):GetButtons(1)) {
            button.setVisibility(View.VISIBLE);
        }
    }

    public boolean KeyboardToolbarModifier(int viewID) {
        if (viewID == R.id.moreOpt) {
            // Place holder to do nothing,
            // on click listener will be done instead and is setup
        } else if (viewID == R.id.winKey) {
            return KeyboardToolbarModifierHandler(!winActive, "WIN(", "Win(", () -> winActive = true);
        } else if (viewID == R.id.fnKey) {
            return KeyboardToolbarModifierHandler(!fnActive, "FN+", null, () -> fnActive = true);
        } else if (viewID == R.id.altKey) {
            return KeyboardToolbarModifierHandler(!altActive, "%(", "Alt(", () -> altActive = true);
        } else if (viewID == R.id.ctrlKey) {
            return KeyboardToolbarModifierHandler(!ctrlActive, "^(", "Ctrl(", () -> ctrlActive = true);
        } else if (viewID == R.id.shiftKey) {
            return KeyboardToolbarModifierHandler(!shiftActive, "+(", "Shift(", () -> shiftActive = true);
        } else {
            return false;
        }
        return true;
    }

    private boolean KeyboardToolbarModifierHandler(boolean condition, String keyComb, String keyboardText, Runnable activationAction) {
        if (condition) {
            activationAction.run();
            keyCombination.append(keyComb);
            if (keyboardText != null) {
                keyboardTextInputView.append(keyboardText);
            }
            parenthesesCount += 1;
            return true;
        }
        return false;
    }

    public void ResetKeyboardModifiers() {
        winActive = false;
        ctrlActive = false;
        shiftActive = false;
        altActive = false;
        fnActive = false;
        parenthesesCount = 0;
        keyCombination.setLength(0);
    }
}

File: DAIRemoteApp/app/src/main/java/com/example/dairemote_app/utils/KeyboardToolbar.kt
package com.example.dairemote_app.utils

import android.view.View
import android.widget.GridLayout
import android.widget.TextView
import androidx.appcompat.widget.Toolbar
import com.example.dairemote_app.R

class KeyboardToolbar internal constructor(
    var moreOpts: View,
    private var modifier1: View,
    private var modifier2: View,
    private var modifier3: View,
    private var modifier4: View,
    private var modifier5: View,
    private val keyboardTextInputView: TextView,
    private val keyboardToolbar: Toolbar,
    private val keyboardExtraBtnsLayout: GridLayout,
    private val p1RowsButtons: Array<TextView>,
    private val p2RowsButtons: Array<TextView>
) {
    private val keyCombination = StringBuilder()
    var winActive = false
    var ctrlActive = false
    var shiftActive = false
    var altActive = false
    var fnActive = false
    private var modifierToggled = false
    private val p1Keys = arrayOf(
        "{INS}",
        "{DEL}",
        "{PRTSC}",
        "{TAB}",
        "{UP}",
        "{ESC}",
        "UP",
        "DOWN",
        "MUTE",
        "{LEFT}",
        "{DOWN}",
        "{RIGHT}"
    )
    private val p2Keys = arrayOf(
        "{F1}",
        "{F2}",
        "{F3}",
        "{F4}",
        "{F5}",
        "{F6}",
        "{F7}",
        "{F8}",
        "{F9}",
        "{F10}",
        "{F11}",
        "{F12}"
    )
    private var parenthesesCount = 0
    private var currentPageIndex = 0

    fun getParenthesesCount(): Int {
        return this.parenthesesCount
    }

    fun getKeyboardToolbar(): Toolbar {
        return this.keyboardToolbar
    }

    private fun getKeyboardLayout(): GridLayout {
        return this.keyboardExtraBtnsLayout
    }

    fun getModifierToggled(): Boolean {
        return this.modifierToggled
    }

    fun setModifierToggled(toggled: Boolean) {
        this.modifierToggled = toggled
    }

    fun getKeys(page: Int): Array<String> {
        if (page == 0) {
            return p1Keys
        }
        return p2Keys
    }

    fun getButtons(page: Int): Array<TextView> {
        if (page == 0) {
            return p1RowsButtons
        }
        return p2RowsButtons
    }

    fun getCurrentToolbarPage(): Int {
        return this.currentPageIndex
    }

    private fun setCurrentToolbarPage(page: Int) {
        this.currentPageIndex = page
    }

    fun nextToolbarPage(): Int {
        setCurrentToolbarPage((getCurrentToolbarPage() + 1) % 2)
        return getCurrentToolbarPage()
    }

    fun getKeyboardTextView(): TextView {
        return this.keyboardTextInputView
    }

    fun getKeyCombination(): StringBuilder {
        return this.keyCombination
    }

    private fun appendKeyCombination(character: String?) {
        keyCombination.append(character)
    }

    fun appendKeyCombination(character: Char) {
        keyCombination.append(character)
    }

    fun addParentheses() {
        if (getParenthesesCount() > 0) {
            for (i in 0 until parenthesesCount) {
                appendKeyCombination(")")
            }
            keyboardTextInputView.text = ""
        }
    }

    fun toggleKeyboardToolbar(open: Boolean) {
        if (open) {
            if (getKeyboardToolbar().visibility != View.VISIBLE) {
                getKeyboardToolbar().visibility = View.VISIBLE
                getKeyboardLayout().visibility = View.VISIBLE
                keyboardExtraSetRowVisibility(getCurrentToolbarPage())
            }
        } else {
            if (getKeyboardToolbar().visibility == View.VISIBLE) {
                getKeyboardToolbar().visibility = View.GONE
                getKeyboardLayout().visibility = View.GONE
            }
        }
    }

    fun keyboardExtraSetRowVisibility(pageIndex: Int) {
        // Hide buttons for the current page
        for (button in if ((pageIndex == 0)) getButtons(1) else getButtons(0)) {
            button.visibility = View.GONE
        }

        // Show buttons for the current page
        for (button in if ((pageIndex == 0)) getButtons(0) else getButtons(1)) {
            button.visibility = View.VISIBLE
        }
    }

    fun keyboardToolbarModifier(viewID: Int): Boolean {
        return when (viewID) {
            R.id.moreOpt -> {
                return true
            }

            R.id.winKey -> {
                winActive = !winActive
                keyboardToolbarModifierHandler(
                    winActive,
                    "WIN(",
                    "Win("
                ) { /* No need for action here */ }
            }

            R.id.fnKey -> {
                fnActive = !fnActive
                keyboardToolbarModifierHandler(
                    fnActive,
                    "",
                    ""
                ) { /* No need for action here */ }
            }

            R.id.altKey -> {
                altActive = !altActive
                keyboardToolbarModifierHandler(
                    altActive,
                    "%(",
                    "Alt("
                ) { /* No need for action here */ }
            }

            R.id.ctrlKey -> {
                ctrlActive = !ctrlActive
                keyboardToolbarModifierHandler(
                    ctrlActive,
                    "^(",
                    "Ctrl("
                ) { /* No need for action here */ }
            }

            R.id.shiftKey -> {
                shiftActive = !shiftActive
                keyboardToolbarModifierHandler(
                    shiftActive,
                    "+(",
                    "Shift("
                ) { /* No need for action here */ }
            }

            else -> false
        }
    }

    private fun keyboardToolbarModifierHandler(
        condition: Boolean,
        keyComb: String,
        keyboardText: String?,
        activationAction: Runnable
    ): Boolean {
        if (condition) {
            activationAction.run()
            keyCombination.append(keyComb)
            if (keyboardText != null) {
                keyboardTextInputView.append(keyboardText)
            }
            parenthesesCount += 1
            return true
        }
        return false
    }

    fun setupButtonClickListeners(listener: View.OnClickListener) {
        // Set click listeners for all buttons
        p1RowsButtons.forEach { button ->
            button.setOnClickListener(listener)
        }

        p2RowsButtons.forEach { button ->
            button.setOnClickListener(listener)
        }

        modifier1.setOnClickListener(listener)
        modifier2.setOnClickListener(listener)
        modifier3.setOnClickListener(listener)
        modifier4.setOnClickListener(listener)
        modifier5.setOnClickListener(listener)
    }

    fun resetKeyboardModifiers() {
        winActive = false
        ctrlActive = false
        shiftActive = false
        altActive = false
        fnActive = false
        setModifierToggled(false)
        parenthesesCount = 0
        keyCombination.setLength(0)
    }
}